File: Python/Aula 03/Complementares/ex3.py
class No:
    def __init__(self, dado):
        self.dado = dado
        self.proximo = None

    def __del__(self):
        if self.proximo is not None:
            del self.proximo

class Lista:
    def __init__(self):
        self.cabeca = None

    @classmethod
    def criar_copia(cls, outra_lista):
        nova_lista = cls()
        if outra_lista.cabeca is None:
            return nova_lista
        
        nova_lista.cabeca = No(outra_lista.cabeca.dado)
        atual_orig = outra_lista.cabeca.proximo
        atual_novo = nova_lista.cabeca
        
        while atual_orig is not None:
            atual_novo.proximo = No(atual_orig.dado)
            atual_novo = atual_novo.proximo
            atual_orig = atual_orig.proximo
            
        return nova_lista

    def inserir(self, dado):
        novo = No(dado)
        novo.proximo = self.cabeca
        self.cabeca = novo

    def __del__(self):
        if self.cabeca is not None:
            del self.cabeca

if __name__ == "__main__":

    minha_lista = Lista()
    minha_lista.inserir(10)
    minha_lista.inserir(20)
    print("Lista criada. Ao finalizar, o destrutor limpará os nós.")

File: C++/Aula 03/Complementares/ex3.cpp
#include <iostream>
#include <stdexcept>
using namespace std;

// ──────────────────────────────────────────────
// Nó da lista
// ──────────────────────────────────────────────
struct No{
    int valor;
    No *proximo;

    No(int v) : valor(v), proximo(nullptr){}

    ~No(){}
};

class Lista{
    private:
        No *cabeca;


        void liberarTudo(){
            No *atual = cabeca;
            while (atual != nullptr){
                No *prox = atual->proximo;
                delete atual;
                atual = prox;
            }
            cabeca = nullptr;
        }

    public:
        Lista() : cabeca(nullptr){
            cout << "[+] Lista criada (vazia)\n";
        }

        Lista(const Lista &outra) : cabeca(nullptr){
            No *src = outra.cabeca;
            No **dst = &cabeca;

            while (src != nullptr){
                *dst = new No(src->valor);
                dst = &((*dst)->proximo);
                src = src->proximo;
            }
        cout << "[+] Lista copiada (deep copy)\n";
        }

        Lista &operator=(Lista outra){
            swap(cabeca, outra.cabeca);
            cout << "[=] Lista atribuída (swap)\n";
            return *this;
        }

        ~Lista(){
            liberarTudo();
            cout << "[-] Lista destruída\n";
        }

        void inserir(int valor){
            No *novo = new No(valor);
            novo->proximo = cabeca;
            cabeca = novo;
        }

        void inserirFim(int valor){
            No *novo = new No(valor);
            if (!cabeca){
                cabeca = novo;
                return;
            }
        
            No *cur = cabeca;
            while (cur->proximo)
                cur = cur->proximo;
            cur->proximo = novo;
        }

        void exibir() const{
            No *cur = cabeca;
            cout << "  Lista: ";
            while (cur){
                cout << cur->valor;
            
                if (cur->proximo)
                    cout << " -> ";
                cur = cur->proximo;
            }
        cout << " -> NULL\n";
        }

        bool vazia() const { return cabeca == nullptr; }
};

int main(){
    cout << "=== Lista original ===\n";
    Lista l1;
    l1.inserirFim(10);
    l1.inserirFim(20);
    l1.inserirFim(30);
    l1.exibir();

    cout << "\n=== Cópia (deep copy) ===\n";
    Lista l2(l1);
    l2.inserirFim(40);
    cout << "l1 após modificar l2:\n";
    l1.exibir();
    cout << "l2:\n";
    l2.exibir();

    cout << "\n=== Teste de escopo — destruição automática ===\n";{
        Lista l3;
        for (int i = 1; i <= 5; ++i)
            l3.inserirFim(i * 100);
        l3.exibir();
        cout << "  [Saindo do escopo — l3 será destruída]\n";
    }

    cout << "\n=== Aviso sobre stack overflow ===\n";
    cout << "  Se o destrutor do No fosse recursivo (delete proximo),\n"
              << "  uma lista com ~100.000 nós causaria stack overflow.\n"
              << "  Esta implementação usa loop iterativo no destrutor da Lista.\n";

    cout << "\n[Saindo do main]\n";
    return 0;
}